test/cpp/end2end/server_early_return_test.cc

```diff
@@ -0,0 +1,233 @@
+ *
+ */
+
+#include <grpc++/channel.h>
+#include <grpc++/client_context.h>
+#include <grpc++/create_channel.h>
+#include <grpc++/security/credentials.h>
+#include <grpc++/security/server_credentials.h>
+#include <grpc++/server.h>
+#include <grpc++/server_builder.h>
+#include <grpc++/server_context.h>
+#include <grpc/grpc.h>
+#include <grpc/support/alloc.h>
+#include <grpc/support/log.h>
+#include <grpc/support/thd.h>
+#include <grpc/support/time.h>
+
+#include "src/proto/grpc/testing/echo.grpc.pb.h"
+#include "test/core/util/port.h"
+#include "test/core/util/test_config.h"
+#include "test/cpp/util/string_ref_helper.h"
+
+#include <gtest/gtest.h>
+
+namespace grpc {
+namespace testing {
+namespace {
+
+const char kServerReturnStatusCode[] = "server_return_status_code";
+const char kServerDelayBeforeReturnUs[] = "server_delay_before_return_us";
+const char kServerReturnAfterNReads[] = "server_return_after_n_reads";
+
+class TestServiceImpl : public ::grpc::testing::EchoTestService::Service {
+ public:
+  // Unused methods are not implemented.
+
+  Status RequestStream(ServerContext* context,
+                       ServerReader<EchoRequest>* reader,
+                       EchoResponse* response) override {
+    int server_return_status_code =
+        GetIntValueFromMetadata(context, kServerReturnStatusCode, 0);
+    int server_delay_before_return_us =
+        GetIntValueFromMetadata(context, kServerDelayBeforeReturnUs, 0);
+    int server_return_after_n_reads =
+        GetIntValueFromMetadata(context, kServerReturnAfterNReads, 0);
+
+    EchoRequest request;
+    while (server_return_after_n_reads--) {
+      EXPECT_TRUE(reader->Read(&request));
+    }
+
+    response->set_message("response msg");
+
+    gpr_sleep_until(gpr_time_add(
+        gpr_now(GPR_CLOCK_MONOTONIC),
+        gpr_time_from_micros(server_delay_before_return_us, GPR_TIMESPAN)));
+
+    return Status(static_cast<StatusCode>(server_return_status_code), "");
+  }
+
+  Status BidiStream(
+      ServerContext* context,
+      ServerReaderWriter<EchoResponse, EchoRequest>* stream) override {
+    int server_return_status_code =
+        GetIntValueFromMetadata(context, kServerReturnStatusCode, 0);
+    int server_delay_before_return_us =
+        GetIntValueFromMetadata(context, kServerDelayBeforeReturnUs, 0);
+    int server_return_after_n_reads =
+        GetIntValueFromMetadata(context, kServerReturnAfterNReads, 0);
+
+    EchoRequest request;
+    EchoResponse response;
+    while (server_return_after_n_reads--) {
+      EXPECT_TRUE(stream->Read(&request));
+      response.set_message(request.message());
+      EXPECT_TRUE(stream->Write(response));
+    }
+
+    gpr_sleep_until(gpr_time_add(
+        gpr_now(GPR_CLOCK_MONOTONIC),
+        gpr_time_from_micros(server_delay_before_return_us, GPR_TIMESPAN)));
+
+    return Status(static_cast<StatusCode>(server_return_status_code), "");
+  }
+
+  int GetIntValueFromMetadata(ServerContext* context, const char* key,
```

Comment:
This is copied from end2end and async_end2end right? Nit, and no need to do it in this PR, but we should create a helper lib with this and other util functions now that we have it in three places